pennylane/devices/qubit/simulate.py

```diff
@@ -237,7 +254,189 @@ def simulate(
     tensor([0.68117888, 0.        , 0.31882112, 0.        ], requires_grad=True))
 
     """
-    state, is_state_batched = get_final_state(circuit, debugger=debugger, interface=interface)
+    has_mcm = has_mid_circuit_measurements(circuit)
+    has_shots = circuit.shots.total_shots is not None
+    if has_mcm and has_shots:
+        if circuit.shots.has_partitioned_shots:
+            results = []
+            for shots_copy in circuit.shots.shot_vector:
+                for _ in range(shots_copy[1]):
+                    aux_circuit = circuit.copy()
+                    aux_circuit._shots = qml.measurements.Shots(shots_copy[0])
+                    results.append(simulate(aux_circuit, rng, prng_key, debugger, interface))
+            return tuple(results)
+        aux_circuit = init_auxiliary_circuit(circuit)
+        one_shot_meas, tmp_dict = simulate_native_mcm(
+            aux_circuit, rng, prng_key, debugger, interface
+        )
+        mcm_shot_meas = [tmp_dict]
+        for _ in range(circuit.shots.total_shots - 1):
+            tmpmeas, tmp_dict = simulate_native_mcm(aux_circuit, rng, prng_key, debugger, interface)
+            one_shot_meas = accumulate_native_mcm(aux_circuit, one_shot_meas, tmpmeas)
+            mcm_shot_meas.append(tmp_dict)
+        return gather_native_mid_circuit_measurements(circuit, one_shot_meas, mcm_shot_meas)
+    state, is_state_batched, _ = get_final_state(circuit, debugger=debugger, interface=interface)
     if state_cache is not None:
         state_cache[circuit.hash] = state
     return measure_final_state(circuit, state, is_state_batched, rng=rng, prng_key=prng_key)
+
+
+def init_auxiliary_circuit(circuit):
```

Comment:
I recommend adding more details about arguments and returns to the docstring of this function, as well as other non-trivial functions that have been added.